pkg/controller/route53hostedzone/zonereconciler.go

```diff
@@ -0,0 +1,336 @@
+		return
+	}
+
+	// Setup any overrides for the default functions. This is useful for dependency injection.
+	if newZoneReconcilerInput.functionOverrides.getDesiredState != nil {
+		zr.funcs.getDesiredState = newZoneReconcilerInput.functionOverrides.getDesiredState
+	}
+
+	if newZoneReconcilerInput.functionOverrides.getCOAWSClient != nil {
+		zr.funcs.getCOAWSClient = newZoneReconcilerInput.functionOverrides.getCOAWSClient
+	}
+
+	if newZoneReconcilerInput.functionOverrides.getCurrentState != nil {
+		zr.funcs.getCurrentState = newZoneReconcilerInput.functionOverrides.getCurrentState
+	}
+
+	if newZoneReconcilerInput.functionOverrides.createRoute53HostedZone != nil {
+		zr.funcs.createRoute53HostedZone = newZoneReconcilerInput.functionOverrides.createRoute53HostedZone
+	}
+
+	if newZoneReconcilerInput.functionOverrides.deleteRoute53HostedZone != nil {
+		zr.funcs.deleteRoute53HostedZone = newZoneReconcilerInput.functionOverrides.deleteRoute53HostedZone
+	}
+
+	if newZoneReconcilerInput.functionOverrides.awsClientBuilder != nil {
+		zr.funcs.awsClientBuilder = newZoneReconcilerInput.functionOverrides.awsClientBuilder
+	}
+}
+
+// getDesiredState gets the current desired state from kubernetes.
+// If ZoneReconciler.desiredState is not nil, an error will be returned.
+func (zr *ZoneReconciler) getDesiredState() error {
+	if zr.desiredState != nil {
+		// This should never happen.
+		return fmt.Errorf("Desired State already populated for %v/%v", zr.namespace, zr.name)
+	}
+
+	dnszone, err := zr.dnsZonesLister.DNSZones(zr.namespace).Get(zr.name)
+	if err != nil {
+		return err
+	}
+
+	zr.desiredState = dnszone.DeepCopy()
+	zr.dnszone = zr.desiredState.Spec.DNSZone
+
+	return nil
+}
+
+// getCOAWSClient generates a cluster operator aws client.
+// If ZoneReconciler.coAWSClient is not nil, an error will be returned.
+func (zr *ZoneReconciler) getCOAWSClient() error {
+	if zr.coAWSClient != nil {
+		// This should never happen.
+		return fmt.Errorf("coAWSClient already populated for %v/%v", zr.namespace, zr.name)
+	}
+
+	// This allows for using host profiles for AWS auth.
+	secretName := ""
+	regionName := ""
+	if zr.desiredState != nil && zr.desiredState.Spec.AWS != nil {
+		secretName = zr.desiredState.Spec.AWS.AccountSecret.Name
+		regionName = zr.desiredState.Spec.AWS.Region
+	}
+
+	coawsclient, err := zr.funcs.awsClientBuilder(zr.kubeClient, secretName, zr.namespace, regionName)
+	if err != nil {
+		zr.logger.Errorf("Error creating COAWSClient: %v", err)
+		return err
+	}
+
+	zr.coAWSClient = coawsclient
+	return nil
+}
+
+// getCurrentState gets the AWS object for the zone.
+// If ZoneReconciler.currentState is not nil, an error will be returned.
+func (zr *ZoneReconciler) getCurrentState() error {
+	if zr.currentState != nil {
+		// This should never happen.
+		return fmt.Errorf("Current State already populated for %v/%v", zr.namespace, zr.name)
+	}
+
+	output, err := zr.coAWSClient.ListHostedZones(&route53.ListHostedZonesInput{})
```

Comment:
If you test this by mocking coAWSClient instead of the function pointer overrides, we'll have a bit better test coverage of all the other logic in here.